src/dataframerow/utils.jl

```diff
@@ -135,14 +129,47 @@ function row_group_slots(cols::Tuple{Vararg{AbstractVector}},
             groups[i] = gix
         end
     end
-    return ngroups, rhashes, gslots
+    return ngroups, rhashes, gslots, false, true
+end
+
+function row_group_slots(cols::Tuple{CategoricalVector},
+                         hash::Val{false},
+                         groups::Union{Vector{Int}, Nothing} = nothing,
+                         skipmissing::Bool = false)::Tuple{Int, Vector{UInt}, Vector{Int}, Bool, Bool}
+    col = cols[1]
+    @assert groups === nothing || length(groups) == length(col)
+
+    # If missings are to be skipped, they will all go to group 1
+    ngroups = length(levels(col)) + (eltype(col) >: Missing)
+
+    if groups !== nothing
+        # When levels are in the same order as the index and there are no missing values,
+        # we could simply copy refs to groups, but the performance gain is negligible,
+        # so always sort groups in the order of levels
+        refmap = [0; CategoricalArrays.order(col.pool)]
+        if skipmissing
+            refmap .+= 1
+        else
+            refmap[1] = ngroups
+        end
+        @inbounds for i in eachindex(groups)
+            groups[i] = refmap[col.refs[i]+1]
+        end
+    end
+    return ngroups, UInt[], Int[], true, true
 end
 
 # Builds RowGroupDict for a given DataFrame.
 # Partly uses the code of Wes McKinney's groupsort_indexer in pandas (file: src/groupby.pyx).
-function group_rows(df::AbstractDataFrame, skipmissing::Bool = false)
+# - hash: whether row hashes should be computed (if false, the rhashes and gslots fields
+#   hold empty vectors)
+# - sort: whether groups should be sorted
+# - skipmissing: whether rows with missing values should be skipped rather than put into group 0
```

Comment:
group `ngroups` not `0` unless I do not understand something.

Reply:
Actually, the group isn't necessarily the first nor the last one if `sort=false`.